pkg/controller/populators/upload-populator.go

```diff
@@ -0,0 +1,231 @@
+package populators
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	cdiv1 "kubevirt.io/containerized-data-importer-api/pkg/apis/core/v1beta1"
+	cc "kubevirt.io/containerized-data-importer/pkg/controller/common"
+	featuregates "kubevirt.io/containerized-data-importer/pkg/feature-gates"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	uploadPopulatorName = "upload-populator"
+
+	// UploadFailed provides a const to indicate upload has failed
+	errUploadFailed = "UploadFailed"
+	// UploadSucceeded provides a const to indicate upload has succeeded
+	uploadSucceeded = "UploadSucceeded"
+
+	// messageUploadFailed provides a const to form upload has failed message
+	messageUploadFailed = "Upload into %s failed"
+	// messageUploadSucceeded provides a const to form upload has succeeded message
+	messageUploadSucceeded = "Successfully uploaded into %s"
+)
+
+// UploadPopulatorReconciler members
+type UploadPopulatorReconciler struct {
+	ReconcilerBase
+}
+
+// NewUploadPopulator creates a new instance of the upload populator
+func NewUploadPopulator(
+	ctx context.Context,
+	mgr manager.Manager,
+	log logr.Logger,
+	installerLabels map[string]string,
+) (controller.Controller, error) {
+	client := mgr.GetClient()
+	reconciler := &UploadPopulatorReconciler{
+		ReconcilerBase: ReconcilerBase{
+			client:          client,
+			scheme:          mgr.GetScheme(),
+			log:             log.WithName(uploadPopulatorName),
+			recorder:        mgr.GetEventRecorderFor(uploadPopulatorName),
+			featureGates:    featuregates.NewFeatureGates(client),
+			installerLabels: installerLabels,
+		},
+	}
+
+	uploadPopulator, err := controller.New(uploadPopulatorName, mgr, controller.Options{
+		Reconciler: reconciler,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	if err := addCommonPopulatorsWatches(mgr, uploadPopulator); err != nil {
+		return nil, err
+	}
+	if err := addUploadPopulatorWatches(mgr, uploadPopulator, log); err != nil {
+		return nil, err
+	}
+	return uploadPopulator, nil
+}
+
+func addUploadPopulatorWatches(mgr manager.Manager, c controller.Controller, log logr.Logger) error {
```

Comment:
I think these watches and indexer could also be included in `addCommonPopulatorsWatches`, maybe just adding a `typeToWatch` argument as we do here https://github.com/kubevirt/containerized-data-importer/blob/main/pkg/controller/datavolume/clone-controller-base.go#L697.